osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs

```diff
@@ -107,9 +113,78 @@ public void AddRotation(float angle)
             }
 
             currentRotation += angle;
+
             // rate has to be applied each frame, because it's not guaranteed to be constant throughout playback
             // (see: ModTimeRamp)
-            drawableSpinner.Result.RateAdjustedRotation += (float)(Math.Abs(angle) * (gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate));
+            double time = gameplayClock?.CurrentTime ?? Clock.CurrentTime;
+            double rate = gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate;
+            float rateAdjustedRotation = (float)(angle * rate);
+
+            if (rate > 0)
+            {
+                int direction = Math.Sign(rateAdjustedRotation);
+
+                if (direction != 0 && direction != currentDirection)
+                {
+                    turningPoints.Add((time, currentDelta));
+                    currentDelta = 0;
+                }
+
+                currentDirection = direction;
+            }
+            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
+            {
+                currentDelta = turningPoints[^1].delta;
+                currentDirection = Math.Sign(currentDelta);
+                turningPoints.RemoveAt(turningPoints.Count - 1);
+            }
```

Comment:
This is going to need some heavy inline commenting.. I'm having difficulty understanding the top level branches, and the importance of the fall-through condition (if that's a valid case).

Reply:
I dunno, seemed pretty straightforward to me when I read it - what's unclear here specifically?

Reply:
Having the `else` condition also include subcondition reduces readability. It's also unclear what anything is doing anywhere. I'd go as far as saying in the current state, with no inline commenting, this reads about on-par with stable's spinner code.

I'll resolve this one and do a more thorough code review pass to add my thoughts.